dali/kernels/imgproc/jpeg/chroma_subsample_gpu.cuh

```diff
@@ -0,0 +1,221 @@
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_JPEG_CHROMA_SUBSAMPLE_GPU_H_
+#define DALI_KERNELS_IMGPROC_JPEG_CHROMA_SUBSAMPLE_GPU_H_
+
+#include <cuda_runtime_api.h>
+#include <cuda_fp16.h>
+
+#include "dali/kernels/common/block_setup.h"
+#include "dali/kernels/imgproc/surface.h"
+#include "dali/kernels/imgproc/sampler.h"
+#include "dali/core/geom/vec.h"
+
+namespace dali {
+namespace kernels {
+
+struct SampleDesc {
+  const uint8_t *in;  // rgb
+  uint8_t *out;  // rgb
+  ivec<2> size;
+  i64vec<2> strides;
+};
+
+template <typename T>
+__inline__ __device__ T rgb_to_y(vec<3, T> rgb) {
+  return ConvertSat<T>(0.299f * rgb.x + 0.587f * rgb.y + 0.114f * rgb.z);
+}
+
+template <typename T>
+__inline__ __device__ T rgb_to_cb(vec<3, T> rgb) {
+  return ConvertSat<T>(-0.16873589f * rgb.x - 0.33126411f * rgb.y + 0.50000000f * rgb.z + 128.0f);
+}
+
+template <typename T>
+__inline__ __device__ T rgb_to_cr(vec<3, T> rgb) {
+  return ConvertSat<T>(0.50000000f * rgb.x - 0.41868759f * rgb.y - 0.08131241f * rgb.z + 128.0f);
+}
+
+template <typename T>
+__inline__ __device__ vec<2, T> rgb_to_cb_cr(vec<3, T> rgb) {
+  return {rgb_to_cb<T>(rgb), rgb_to_cr<T>(rgb)};
+}
+
+template <int N, typename T>
+__inline__ __device__ vec<N> avg4(vec<N, T> a, vec<N, T> b, vec<N, T> c, vec<N, T> d) {
+  IMPL_VEC_ELEMENTWISE((a[i] + b[i] + c[i] + d[i]) * 0.25f);
+}
+
+template <int N, typename T>
+__inline__ __device__ vec<N> avg2(vec<N, T> a, vec<N, T> b) {
+  IMPL_VEC_ELEMENTWISE((a[i] + b[i]) * 0.5f);
+}
+
+template <typename T>
+__inline__ __device__ vec<3, T> ycbcr_to_rgb(const vec<3, T> ycbcr) {
+    float y  = static_cast<float>(ycbcr.x);
+    float cb = static_cast<float>(ycbcr.y) - 128.0f;
+    float cr = static_cast<float>(ycbcr.z) - 128.0f;
+    vec<3, T> rgb;
+    rgb.x = ConvertSat<T>(y + 1.402f * cr);
+    rgb.y = ConvertSat<T>(y - 0.34413629f * cb - 0.71413629f * cr);
+    rgb.z = ConvertSat<T>(y + 1.772f * cb);
+    return rgb;
+}
+
+template <bool horz_subsample, bool vert_subsample, typename T>
+__inline__ __device__ 
+void rgb_to_ycbcr_chroma_subsample(int local_x, int local_y,
+                                   const Surface2D<T>& luma,
+                                   const Surface2D<T>& cb,
+                                   const Surface2D<T>& cr,
+                                   const Surface2D<const uint8_t>& in) {
+  const auto sampler = make_sampler<DALI_INTERP_NN>(in);
+  int chroma_y = local_y;
+  int chroma_x = local_x;
+  int y = chroma_y << vert_subsample;
+  int x = chroma_x << horz_subsample;
+
+  vec<3, T> rgb[4];
+  sampler(rgb[0].v, ivec2(x, y), BorderClamp());
+  luma(x, y) = rgb_to_y<T>(rgb[0]);
+  vec<3, T> avg_rgb(rgb[0]);
+  if (horz_subsample && vert_subsample) {
+    sampler(rgb[1].v, ivec2(x + 1, y), BorderClamp());
+    sampler(rgb[2].v, ivec2(x, y + 1), BorderClamp());
+    sampler(rgb[3].v, ivec2(x + 1, y + 1), BorderClamp());
+    avg_rgb = avg4(rgb[0], rgb[1], rgb[2], rgb[3]);
```

Comment:
Maybe you can move `avg_rgb` to the end just to be consistent with the other branches: sampler, luma, avg.